### PR TITLE
Feature/sched 26

### PR DESCRIPTION
**Pull request by stuemke**
new feature to record job executions.

### CODE REVIEW

File: src/main/java/org/appng/application/scheduler/job/JobRecordHouseKeepingJob.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+import org.appng.api.model.Application;
+import org.appng.api.model.Site;
+import org.appng.application.scheduler.service.JobRecordService;
+
+/**
+ * A {@link ScheduledJob} to remove old job records from the database.
+ * 
+ * @author Claus Stümke
+ *
+ */
```
Comment (madness-inc) [inline src/main/java/org/appng/application/scheduler/job/JobRecordHouseKeepingJob.java:34]: Why not use lombok's @Data?

File: src/main/java/org/appng/application/scheduler/service/RecordsRestController.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+/**
+ * {@link RestController} providing a REST API to query for job records.
+ * 
+ * @author Claus Stümke
+ *
+ */
+@RestController
+public class RecordsRestController {
+
+	@Autowired
```
Comment (madness-inc) [inline src/main/java/org/appng/application/scheduler/service/RecordsRestController.java]: use constructor injection, makes getter/setter obsolete

File: src/main/java/org/appng/application/scheduler/model/JobResult.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.appng.application.scheduler.model;
+
+import org.appng.api.ScheduledJobResult;
+import org.appng.api.ScheduledJobResult.ExecutionResult;
+
+import com.fasterxml.jackson.annotation.JsonIgnore;
+
```
Comment (madness-inc) [inline src/main/java/org/appng/application/scheduler/model/JobResult.java:22]: Wyh not use @Data?

File: src/main/java/org/appng/application/scheduler/model/JobRecord.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+import com.fasterxml.jackson.annotation.JsonFormat;
+
+/**
+ * Model class representig a row in the job record table.
+ * 
+ * @author Claus Stümke
+ *
+ */
+
```
Comment (madness-inc) [inline src/main/java/org/appng/application/scheduler/model/JobRecord.java]: Why not use @Data?

File: application-home/beans.xml [hunk 5]
```diff
@@ -79,11 +95,32 @@
...
 		</property>
 	</bean>
+	
+	
+	<bean id="houseKeepingJob" class="org.appng.application.scheduler.job.JobRecordHouseKeepingJob">
+		<property name="description" value="Job Record House Keeping Job" />
+		<property name="jobDataMap">
+			<map>
+				<entry key="enabled" value="${houseKeepingEnabled}" />
+				<entry key="cronExpression" value="${indexExpression}" />
```
Comment (madness-inc) [inline application-home/beans.xml]: indexExpression is used for lucene indexing. Use dedicated property?

File: src/main/java/org/appng/application/scheduler/business/Records.java [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+
+import com.google.common.collect.Lists;
+
+/**
+ * A {@link DataProvider} for recorded job executions. It supports some filters.
+ * 
+ * @author Claus Stümke
+ *
+ */
+public class Records implements DataProvider {
```
Comment (madness-inc) [inline src/main/java/org/appng/application/scheduler/business/Records.java:49]: Define as @Component and use constructor injection (remove from beans.xml)

File: src/main/java/org/appng/application/scheduler/business/SchedulingController.java [hunk 7]
```diff
@@ -36,16 +37,20 @@
 import org.quartz.SchedulerException;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.data.domain.Pageable;
 
 public class SchedulingController extends SchedulerAware implements ApplicationController {
```
Comment (madness-inc) [inline src/main/java/org/appng/application/scheduler/business/SchedulingController.java:45]: Define as @Component and use constructor injection (remove from beans.xml)

File: src/main/java/org/appng/application/scheduler/business/Records.java [hunk 8]
```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright 2011-2017 the original author or authors.
```
Comment (madness-inc) [inline src/main/java/org/appng/application/scheduler/business/Records.java]: must be 2011-2018

**Review by madness-inc — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/org/appng/application/scheduler/business/Records.java (hunk 6 above)
Comment (stuemke) [reply src/main/java/org/appng/application/scheduler/business/Records.java:49]: done. see latest commit

File: src/main/java/org/appng/application/scheduler/business/SchedulingController.java (hunk 7 above)
Comment (stuemke) [reply src/main/java/org/appng/application/scheduler/business/SchedulingController.java:45]: done. see latest commit

File: src/main/java/org/appng/application/scheduler/job/JobRecordHouseKeepingJob.java (hunk 1 above)
Comment (stuemke) [reply src/main/java/org/appng/application/scheduler/job/JobRecordHouseKeepingJob.java:34]: done. see latest commit

File: src/main/java/org/appng/application/scheduler/model/JobResult.java (hunk 3 above)
Comment (stuemke) [reply src/main/java/org/appng/application/scheduler/model/JobResult.java:22]: done. see latest commit. But you have to be carefule using @ Annotations in comments. Maybe you can change the one above?

File: src/main/java/org/appng/application/scheduler/service/JobRecordService.java [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+				record.setScheduledJobResult(new ScheduledJobResult());
+				record.getScheduledJobResult().setResult(ExecutionResult.valueOf(rs.getString(FIELD_NAME_RESULT)));
+				record.setStacktraces(rs.getString(FIELD_NAME_STACKTRACES));
+				return record;
+			}
+		});
+
+	}
+
+	private boolean addFiler(String filterName, String operator, String filterValue, StringBuilder sql,
```
Comment (madness-inc) [inline src/main/java/org/appng/application/scheduler/service/JobRecordService.java]: Shoud be addFil**t**er

File: src/main/java/org/appng/application/scheduler/service/JobRecordService.java [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+			MapSqlParameterSource paramsMap, boolean first) {
+		return addFiler(filterName, operator, filterValue, sql, paramsMap, first, filterName);
+	}
+
+	private boolean addFiler(String filterName, String operator, String filterValue, StringBuilder sql,
```
Comment (madness-inc) [inline src/main/java/org/appng/application/scheduler/service/JobRecordService.java:151]: Shoud be addFil**t**er

File: src/main/java/org/appng/application/scheduler/business/Records.java [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+				START_FILTER_DATE_TIME_FORMAT);
+		filter.getSelections().add(startedAfterFilter);
+
+		Selection startedBeforeFilter = selectionFactory.getDateSelection(START_BEFORE_FILTER,
+				MessageConstants.FILTER_RECORD_STARTED_BEFORE, request.getParameter(START_BEFORE_FILTER),
+				START_FILTER_DATE_TIME_FORMAT);
+		filter.getSelections().add(startedBeforeFilter);
+	}
+
+	public JobRecordService getJobRecordService() {
```
Comment (madness-inc) [inline src/main/java/org/appng/application/scheduler/business/Records.java]: Getter/setter not need due to constructor injection

Comment (stuemke) [reply src/main/java/org/appng/application/scheduler/business/Records.java]: getter and setter have been removed

File: src/main/java/org/appng/application/scheduler/service/JobRecordService.java (hunk 10 above)
Comment (stuemke) [reply src/main/java/org/appng/application/scheduler/service/JobRecordService.java:151]: Typo has been corrected

File: pom.xml [hunk 12]
```diff
@@ -156,6 +173,47 @@
 					</execution>
 				</executions>
 			</plugin>
+			<plugin>
```
Comment (madness-inc) [inline pom.xml:178]: Use appNG Maven Plugin instead
https://github.com/appNG/appng/tree/appng-1.17.1/appng-maven-plugin

Comment (stuemke) [reply pom.xml:178]: Done. Using appng-maven-plugin now

**Review by madness-inc — APPROVED** [review]
Great work!

**Review by madness-inc — APPROVED** [review]
Great Work!

**Pull request merged**